narwhal/primary/src/synchronizer.rs

```diff
@@ -1,80 +1,147 @@
+        worker_cache: SharedWorkerCache,
         certificate_store: CertificateStore,
         payload_store: Store<(BatchDigest, WorkerId), PayloadToken>,
-        tx_header_waiter: Sender<WaiterMessage>,
         tx_certificate_waiter: Sender<Certificate>,
+        rx_consensus_round_updates: watch::Receiver<Round>,
         dag: Option<Arc<Dag>>,
     ) -> Self {
-        let mut synchronizer = Self {
+        let genesis = Self::make_genesis(&committee.load());
+        Self {
             name,
+            committee,
+            worker_cache,
             certificate_store,
             payload_store,
-            tx_header_waiter,
             tx_certificate_waiter,
-            genesis: Vec::default(),
+            rx_consensus_round_updates,
+            genesis: Arc::new(ArcSwap::from_pointee(genesis)),
             dag,
-        };
-        synchronizer.update_genesis(committee);
-        synchronizer
+        }
     }
 
-    /// Update the genesis (called upon reconfiguration).
-    pub fn update_genesis(&mut self, committee: &Committee) {
-        self.genesis = Certificate::genesis(committee)
-            .into_iter()
-            .map(|x| (x.digest(), x))
-            .collect();
+    fn make_genesis(committee: &Committee) -> (Epoch, Vec<(CertificateDigest, Certificate)>) {
+        (
+            committee.epoch(),
+            Certificate::genesis(committee)
+                .into_iter()
+                .map(|x| (x.digest(), x))
+                .collect(),
+        )
     }
 
-    /// Returns `true` if we have all transactions of the payload. If we don't, we return false,
-    /// synchronize with other nodes (through our workers), and re-schedule processing of the
-    /// header for when we will have its complete payload.
-    pub async fn missing_payload(&mut self, header: &Header) -> DagResult<bool> {
-        // We don't store the payload of our own workers.
+    /// Returns genesis certificates for the given Epoch, or returns error if
+    /// the Epoch is not current.
+    fn genesis_for_epoch(
+        &self,
+        epoch: Epoch,
+    ) -> DagResult<Guard<Arc<(Epoch, Vec<(CertificateDigest, Certificate)>)>>> {
+        let genesis_guard = self.genesis.load();
+        match genesis_guard.0.cmp(&epoch) {
+            Ordering::Less => {
+                // Attempt to update cached genesis certs.
+                let committee = self.committee.load();
+                if committee.epoch() != epoch {
+                    debug!(
+                        "synchronizer unable to load a new enough committee: needed {epoch} but got {}",
+                        committee.epoch()
+                    );
+                    return Err(DagError::InvalidEpoch {
+                        expected: committee.epoch(),
+                        received: epoch,
+                    });
+                }
+                self.genesis.store(Arc::new(Self::make_genesis(&committee)));
+                self.genesis_for_epoch(epoch)
```

Comment:
Why the recursion here? Couldn't we just return?

Reply:
No, because we need to return the Guard type